### PR TITLE
FN-7039: graph-native workflow steps (U1–U8)

### PR DESCRIPTION
**Pull request by gsxdsm**
## FN-7039: graph-native workflow steps (complete — U1–U8)

Fixes the FN-7039 report — *optional workflow steps were marked done without running, and didn't show in the step progress bar* — and migrates workflow steps **entirely** onto the workflow-graph machinery, removing the legacy parallel system end-to-end. Plan: `docs/plans/2026-06-25-001-refactor-workflow-steps-graph-native-plan.md`.

### Root cause
`Store.optionalGroupIdSet()` returned an empty set when a task had no explicit `workflowId` and the project had no `defaultWorkflowId`, so a built-in group id (`browser-verification`) collided with a template id and was materialized into a legacy `WS-xxx` row the graph executor never matched. Graph-run steps also never recorded results, so the progress bar stayed empty.

### What changed
- **U1** — `optionalGroupIdSet` → `builtin:coding` fallback; create-time toggles resolve `builtin:coding`. Enabled built-in group ids no longer downgraded.
- **U2** — the graph executor records each enabled optional-group step into `task.workflowStepResults` (keyed by node id) + emits `[pre-merge]` logs.
- **U3** — progress bar + Workflow tab read graph-written results; dropped `workflowStepNameLookup`; added running / advisory-vs-blocking render states.
- **U4** — deleted the legacy `runWorkflowSteps` execution path + `workflow-step` seam/primitive; watchdog re-enters the graph; store-fallback fails closed.
- **U5** — removed the `/api/workflow-steps*` REST surface + Settings → Workflow Steps UI.
- **U6** — deleted `WORKFLOW_STEP_TEMPLATES` (inlined into the IR builders); plugin templates remain as the editor palette.
- **U7** — **post-merge made graph-native and the legacy table removed**:
  - **U7a** post-merge steps run as graph nodes after `merge-attempt` (flag-gated; built-in `coding` byte-identical when off).
  - **U7b** `graphNativePostMerge` default-ON; merger post-merge path made inert (no double-run, proven); migration 130 normalizes built-in enable ids (`WS-004` → `browser-verification`). Verified against real DBs: **no custom/plugin post-merge steps exist**; the only post-merge step (compound-engineering's `document`) already runs as a graph node.
  - **U7c** removed every remaining `workflow_steps` reader/writer (merger post-merge code, store CRUD, `materializeWorkflowSteps`, executor recovery read) and **dropped the table** (migration 131). `materializeWorkflowSteps` was proven vestigial (the graph runs IR nodes, never compiled rows).
- **U8** — docs rewritten to the graph-native model; `settings-reference` updated.
- **Review fixes** — dead-code cleanup, step output/notes recording, engine/core seam-contract sync.

### Verification
- `pnpm test:gate` green (engine-core 298, ci-shape 58); **`pnpm smoke:boot` PASS** (app boots + `/api/health` 200 with the table gone); full `@fusion/core` suite (6290) + reliability backstop (154) green; core/engine/dashboard typecheck clean; seed-at-129 and seed-at-130 migration tests green.
- 4-persona code review (correctness/adversarial/reliability/maintainability) on U1–U6; `safe_auto` + on-goal fixes applied.

### Decision provenance
A six-reviewer doc review changed two foundations (see the plan revision note): the graph writes the **existing `workflowStepResults` field** (no new table — avoids upgrade data loss), and "drop the table" was found unsafe until post-merge went graph-native (the user opted to do the full post-merge rebuild + drop in this PR).

## Residual Review Findings (follow-up)
- **[P2] FN-4343 per-step scope gate not replicated on the graph path** — the per-step `workflowStepScopeEnforcement` leak check lived only in the deleted `runWorkflowSteps`. Merge-time `FileScopeViolationError` + squash-overlap are unaffected; a coding-capable user-authored step could write off-scope files that ride along to merge. Re-introduce post-node scope enforcement on the graph path.
- **[P2] Recording for non-optional-group / split-branch step realizations** — `recordWorkflowStepResult` fires only for `optional-group` nodes in the main traversal; built-in gates record correctly.
- **[P3] Misleading recovery log** — `recoverCompletedTask` logs success-flavored "Auto-recovered…" even when the graph parked the task `failed` (observability only).
- **[P3] Malformed-advisory verdict maps to `passed`** rather than `advisory_failure`. Edge case.
- **[testing] Fail-closed store-capability test** — assert all production store constructions expose `getTaskWorkflowSelection`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Optional workflow steps now consistently run and appear in task progress, including when no project default workflow is set.
  * Task progress now derives workflow step names and statuses from recorded results, showing correct **running** and non-blocking **advisory-failure** behavior, while excluding disabled steps from totals.
* **New Features**
  * Post-merge optional steps now run **graph-native by default** and remain non-blocking on revise outcomes.
* **Removals**
  * Removed legacy workflow-step CRUD/refinement flows and the built-in template-catalog-based model; workflow steps now rely on the graph-native optional-group approach.

### CODE REVIEW

**Comment by ghost** [comment]
Ready to review this PR? Stage has broken it down into 10 individual chapters for you:

| | Title |
|---|-------|
| 1 | [Fix root-cause ID downgrade and fallback](https://stagereview.app/Runfusion/Fusion/pull/1788/chapters/1) |
| 2 | [Enable graph-native workflow step recording](https://stagereview.app/Runfusion/Fusion/pull/1788/chapters/2) |
| 3 | [Update UI to read graph-written results](https://stagereview.app/Runfusion/Fusion/pull/1788/chapters/3) |
| 4 | [Delete legacy runWorkflowSteps path](https://stagereview.app/Runfusion/Fusion/pull/1788/chapters/4) |
| 5 | [Remove legacy workflow steps REST API](https://stagereview.app/Runfusion/Fusion/pull/1788/chapters/5) |
| 6 | [Delete built-in step templates catalog](https://stagereview.app/Runfusion/Fusion/pull/1788/chapters/6) |
| 7 | [Migrate post-merge to graph-native](https://stagereview.app/Runfusion/Fusion/pull/1788/chapters/7) |
| 8 | [Normalize enable IDs and drop table](https://stagereview.app/Runfusion/Fusion/pull/1788/chapters/8) |
| 9 | [Clean up UI and IR builders](https://stagereview.app/Runfusion/Fusion/pull/1788/chapters/9) |
| 10 | [Update documentation and ship metadata](https://stagereview.app/Runfusion/Fusion/pull/1788/chapters/10) |

<a href="https://stagereview.app/Runfusion/Fusion/pull/1788">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

<sub>Chapters generated by <a href="https://stagereview.app">Stage</a> for commit 01f198483b3d357d1cdd52ad56cd5cc31b1bae89 on Jun 26, 2026 4:12pm UTC.</sub>

File: .changeset/fn-7039-workflow-steps-graph-native.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+---
+"@runfusion/fusion": minor
+---
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply .changeset/fn-7039-workflow-steps-graph-native.md:3]: Fixed differently — aligned `category: fix` → `category: feature` to match the `minor` bump (adds new graph-native behavior + removes the /api/workflow-steps route surface), rather than downgrading to `patch`.

File: docs/plans/2026-06-25-001-refactor-workflow-steps-graph-native-plan.md [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+- **DB states:** fresh DB vs upgraded (seed-at-129); built-in vs custom vs plugin-originated legacy rows.
+- **Workflow modes:** `builtin:coding`, `builtin:stepwise-coding`, custom workflows with/without optional groups; stores lacking `getTaskWorkflowSelection` (fail-closed).
+
+---
+
+## Symptom Verification (FN-5893)
+
+- **Original symptom:** A task with `browser-verification` enabled (no project default workflow) is marked done without the step running, and enabled steps don't appear in the progress bar (FN-7039).
+- **Exact reproduction:** Create a task, no `workflowId`, no project `defaultWorkflowId`, `enabledWorkflowSteps: ["browser-verification"]`; run through the executor; inspect stored enable ids, logs, `workflowStepResults`, and the unified progress model.
+- **Assertion it is gone:** Enable id stays `browser-verification` (U1, done); the graph runs the step; a `workflowStepResults` entry keyed by node id is recorded with a terminal status (U2); a `[pre-merge]` log is emitted; `getUnifiedTaskProgress` shows the step with the recorded status across desktop + mobile (U3).
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply docs/plans/2026-06-25-001-refactor-workflow-steps-graph-native-plan.md:252]: Not a defect — FN-5893 is the standing-rule id for the Symptom Verification convention (AGENTS.md "Fix the Invariant, Not the Repro (FN-5893)"), deliberately cited in the heading; the task FN-7039 is named in the section body. Leaving as-is.

File: packages/dashboard/app/components/__tests__/TaskCard.test.tsx [hunk 3]
```diff
@@ -1909,15 +1909,18 @@ describe("TaskCard", () => {
     expect(screen.queryByText("1 steps")).toBeNull();
   });
 
-  it("renders workflow checks after normal steps with mapped statuses and phase badges", () => {
+  // FNXC:WorkflowSteps 2026-06-25 — graph-written results drive the card progress; names come from
+  // result.workflowStepName (with raw-id fallback), and advisory_failure (amber) is visually distinct
+  // from failed (red). No board-level name lookup is involved.
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/dashboard/app/components/__tests__/TaskCard.test.tsx]: Fixed — FNXC header date `2026-06-25` → `2026-06-25-00:00` to match the required `yyyy-MM-dd-hh:mm` format.

File: packages/dashboard/app/utils/__tests__/taskProgress.test.ts [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+import { getUnifiedTaskProgress } from "../taskProgress";
+
+/*
+FNXC:WorkflowSteps 2026-06-25 — graph-native progress model (plan U3).
+These tests pin the render-state contract that the progress bar / Workflow tab rely on:
+- names resolve from result.workflowStepName (no DB-row name lookup), with a raw-id fallback;
+- a "pending" result with a startedAt and no completedAt is the `running` state, vs bare `pending`;
+- advisory_failure (non-blocking) is distinct from failed (blocking) and counts as completed;
+- disabled optional steps (absent from enabledWorkflowSteps) never appear in the counter/bar.
+*/
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/dashboard/app/utils/__tests__/taskProgress.test.ts:12]: Fixed — FNXC header date `2026-06-25` → `2026-06-25-00:00` to satisfy the `yyyy-MM-dd-hh:mm` convention.

File: packages/engine/src/workflow-graph-task-runner.ts [hunk 5]
```diff
@@ -233,6 +235,7 @@ export class WorkflowGraphTaskRunner {
         semaphoreAvailability: this.deps.semaphoreAvailability,
         resumeReconcile: this.deps.resumeReconcile,
         logTaskEntry: this.deps.logTaskEntry,
+        recordWorkflowStepResult: this.deps.recordWorkflowStepResult,
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/engine/src/workflow-graph-task-runner.ts:238]: Reviewed — intentional best-effort visibility. The sink upserts idempotently by node id (pending→terminal), swallows all errors so it never affects the run, and the next run re-upserts; no legacy post-merge fallback remains (removed in U7c). A stranded `pending` after an abort is cosmetic, not corruption.

File: packages/core/src/__tests__/architecture-schema-compat.test.ts [hunk 6]
```diff
@@ -78,6 +78,17 @@ describe("architecture schema compatibility", () => {
...
+    // FNXC:WorkflowStepCRUD 2026-06-26-14:00: TRANSIENT migration tables — created by a
+    // historical migration and DROPPED by a later one (e.g. `workflow_steps`, created in
+    // migration 16, dropped in migration 131) — never reach the final schema, so they must
+    // NOT be in SCHEMA_SQL or MIGRATION_ONLY_TABLE_SCHEMAS (which would resurrect them via
+    // ensureSchemaCompatibility). Exclude any table that has a `DROP TABLE` in db.ts.
+    const droppedTables = new Set<string>();
+    for (const match of source.matchAll(/DROP TABLE\s+(?:IF EXISTS\s+)?([A-Za-z_][A-Za-z0-9_]*)/g)) {
+      droppedTables.add(match[1]);
+    }
+    for (const dropped of droppedTables) discoveredTables.delete(dropped);
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/core/src/__tests__/architecture-schema-compat.test.ts:90]: Reviewed — not over-broad. The only tables both CREATE'd and DROP'd are `workflow_steps` (mig 16→132) and `agentLogEntries` (mig 40→102), both transient migration tables intentionally absent from the final schema. Other DROP targets are virtual/FTS tables (not matched by the CREATE TABLE regex) or have no CREATE TABLE — the exclusion changes nothing for them.

File: packages/core/src/experimental-features.ts [hunk 7]
```diff
@@ -11,11 +11,38 @@ workflowGraphExecutor and workflowColumns graduated from Experimental. Runtime g
...
+as before: the graph routes merge-attempt success straight to `end` and the merger still
+owns post-merge steps from the legacy table — zero behavior change, byte-identical
+builtin:coding traversal. When ON, the graph executor lets traversal continue past a
+SUCCESSFUL merge to any post-merge optional-group node reachable from the merge region,
+running it via the same optional-group execution+recording path (phase:"post-merge",
+non-blocking failures). This unit is additive + reversible: a later unit removes the
+legacy merger post-merge path. Mirrors the WORKFLOW_INTERPRETER_DUAL_OBSERVE_FLAG read
+plumbing (named constant + `isExperimentalFeatureEnabled`).
+*/
+export const GRAPH_NATIVE_POST_MERGE_FLAG = "graphNativePostMerge" as const;
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/core/src/experimental-features.ts:45]: Fixed — rewrote the comment: `graphNativePostMerge` is default-ON, explicit `false` opts out, and the legacy merger post-merge path was deleted in U7c (no legacy-table fallback).

File: packages/engine/src/__tests__/workflow-graph-post-merge.test.ts [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+/*
+FNXC:WorkflowPostMerge 2026-06-26-09:00:
+Graph-native post-merge steps (U7 spike). A post-merge optional-group node wired off
+`merge-attempt` success must, WITH the `graphNativePostMerge` flag ON, run AFTER the
+merge seam and record a WorkflowStepResult with phase:"post-merge". WITH the flag OFF
+the merge region stays collapsed and the post-merge node is never reached — it records
+nothing via the graph (the legacy merger still owns post-merge). Post-merge failures are
+non-blocking: the run still completes with the merge-success outcome.
+*/
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/engine/src/__tests__/workflow-graph-post-merge.test.ts:18]: Fixed — comment now says the explicit opt-out (`graphNativePostMerge:false`) disables graph-native post-merge for that run, with no legacy merger fallback (removed in U7c).

File: packages/engine/src/workflow-graph-executor.ts [hunk 9]
```diff
@@ -290,6 +303,42 @@ export class WorkflowGraphExecutor {
...
+     * such that an edge leaves a merge-region node to `t`, where `t` is itself NOT a
+     * merge-region node and NOT `end`, the edge is not a rework back-edge, and the edge
+     * routes on success (no condition or `condition: "success"`). For `builtin:coding`
+     * this set is EMPTY (every merge-region exit goes to another merge-region node or
+     * `end`), so flag-ON is byte-identical to flag-OFF there — the parity oracle holds.
+     * When the flag is OFF the set is left empty and the post-merge hop is never taken,
+     * so existing merge routing (transient→retry, manual hold, branch-group
+     * integration/promotion, recovery-router, failure paths) is wholly unchanged.
+     */
+    const postMergeEnabled = isExperimentalFeatureEnabled(settings, GRAPH_NATIVE_POST_MERGE_FLAG);
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/engine/src/workflow-graph-executor.ts:327]: Fixed — comment updated to default-ON with explicit `false` as the opt-out, matching `DEFAULT_ON_EXPERIMENTAL_FEATURES`.

File: packages/engine/src/workflow-graph-executor.ts [hunk 10]
```diff
@@ -626,6 +757,25 @@ export class WorkflowGraphExecutor {
...
+           * the merge region stays exactly as collapsed before.
+           */
+          for (const entryId of postMergeEntryNodeIds) {
+            const postMerge = await walk(entryId);
+            // A post-merge entry node is never an enclosing rework head, so a
+            // ReworkSignal here would be malformed IR; ignore it rather than bubble a
+            // rework loop out of the merge boundary. Result is intentionally discarded
+            // (non-blocking).
+            void postMerge;
+          }
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/engine/src/workflow-graph-executor.ts:795]: Fixed — each post-merge `walk(entryId)` is now wrapped in try/catch; a throw is logged via the task log sink and traversal continues to the next entry, so a merged task can never be flipped to an executor failure.
